accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1326,7 +1326,7 @@ fresh, non-late, final variable `v` is created. An initializing formal
 argument passed to this formal. An initializer list element of the
 form `id = e` or `this.id = e` is evaluated by evaluating `e` to an
 object `o` and binding `v` to `o`.  During the execution of the
-constructor body, `this` and `id` are bound to the value of `v`.  The
+constructor body, `this` is bound to the value of `v`.  The
```

Comment:
The `id` is referenced normally using the scope or implicit `this.id`.
That makes a difference for something like:
```dart
extension type E(Object? id) {
  E(Object? base, Object? id) : if = base {
    assert(identical(this.id, id));
  };
}
```

Reply:
Sure, but it is still not working as intended if `id` evaluates to an arbitrary object, I'd expect `identical(this, id)` to be true.

Reply:
Would you assume `identical(this.id, id)` to be true in the example above, even if the two arguments are not the same object?
(I'm using `assert`, maybe that's confusing, I'm not claiming it *must* be true, I'm just doing some operation that can be *not* true. Should have used `!identical` I guess.)

So more precisely:
```dart
extension type E._(Object? id) {
  /// The second argument must not be the same as the first.
  E(Object? base, Object? id) : id = base {
    assert(!identical(this.id, id));
  };
}
```
I expect this extension type to work such that `E(2, 3)` is successful, doesn't fail the assert because it compares `2` and `3`, and ends up having `2` as representation value.

I expect that because I expect the same for a similar class:
```dart
class C {
  final Object? id;
  C._(this.id);
  
  /// The second argument must not be the same as the first.
  C(Object? base, Object? id) : id = base {
    assert(!identical(this.id, id))
  };
}
```
That works as I described above.
And I want the extension type to work the same.

Which means that the claim that "... ann `id` are bound to the value of `v`" during the execution of the constructor body is incorrect. It's not in this case, where it's bound to the value of the second argument, not `v` which is the value of the first argument.

(If the claim should be interpreted in any other way than that the identifier `id` in the lexical scope available to the constructor body is bound to the value `v`, then it should be rephrased, because that's the only reading I can see.)

Every instance method and generative constructor body (anything which can access `this`) can access the representation variable as if it was an instance getter on `this`. It's not added to the scope of those members, like this text says, it's added to the member scope of the extension type declaration itself.
Then it's looked up as normal for a `this` reference.

So the "and `id` are" shouldn't be here, because the only thing I can read it as saying, is not correct.

Reply:
Sorry, I missed the point about multiple declarations with the same name.

I think the following change that I mentioned would suffice:

> During the execution of the constructor body, `this` is bound to the value of `v`, and so is the representation variable.

Reply:
Probably. 

But we'll also have to defined "representation variable" then, because (unless my browser's search function is busted), the phrase "representation variable" only occurs _once_ in this document, where we say that it can be promoted.

Reply:
The representation variable is implied in many locations (and we should probably adjust the text to make it explicit in many of those situations, because that's needed in order to avoid the imprecision about shadowing).

For example, we have a couple of occurrences of text like this:

> `this` and the name of the representation are bound as with the getter invocation

which is used to specify the dynamic semantics of an extension type member invocation/tear-off, and we shouldn't say that "the name" is bound to anything: It's the representation variable which is bound to said receiver object, and the name of the representation is an identifier which may or may not resolve to the representation variable in any given expression. We don't need any special rules for that, name resolution is determined by the syntactic structure (nested scopes) and by implicit insertion of `this.`, and I think it would be a significant source of complexity if we introduce a different set of rules about extension types than the ones that we're using for classes/mixins/enums/etc.

Another example is commentary saying "the static analysis considers the representation as a final instance variable".

Another example is "The static type of the representation name is the representation type" which should again specify the static type of the representation variable because the name may resolve to other declarations.

By the way, I noticed at least one occurrence of 'representation\nvariable'.

I agree, we do have to make the representation variable more explicit in a number of locations in the feature specification.